ipa-core/src/protocol/ipa_prf/oprf_padding/mod.rs

```diff
@@ -1,5 +1,814 @@
+        let _ = send_channel
+            .send(
+                RecordId::FIRST,
+                BA32::truncate_from(u128::from(total_number_of_fake_rows)),
+            )
+            .await;
+    }
+    if ctx.role() == h_i_plus_one {
+        let send_channel = send_ctx.send_channel::<BA32>(send_ctx.role().peer(Direction::Right));
+        let _ = send_channel
+            .send(
+                RecordId::FIRST,
+                BA32::truncate_from(u128::from(total_number_of_fake_rows)),
+            )
+            .await;
+    }
+    if ctx.role() == h_out {
+        // receive `total_number_of_fake_rows` from both other helpers and make sure they are the same
+        let recv_channel_right =
+            send_ctx.recv_channel::<BA32>(send_ctx.role().peer(Direction::Right));
+        let from_right = match recv_channel_right.receive(RecordId::FIRST).await {
+            Ok(v) => u32::try_from(v.as_u128()).unwrap(),
+            Err(e) => return Err(e.into()),
+        };
+
+        let recv_channel_left =
+            send_ctx.recv_channel::<BA32>(send_ctx.role().peer(Direction::Left));
+        let from_left = match recv_channel_left.receive(RecordId::FIRST).await {
+            Ok(v) => u32::try_from(v.as_u128()).unwrap(),
+            Err(e) => return Err(e.into()),
+        };
+        assert_eq!(from_right, from_left);
+        total_number_of_fake_rows = from_right;
+    }
+
+    // Step 3: `h_out` will generate secret shares of zero for as many rows as the `total_number_of_fake_rows`
+    if ctx.role() == h_out {
+        for _ in 0..total_number_of_fake_rows as usize {
+            let row = OPRFIPAInputRow {
+                match_key: Replicated::new(BA64::ZERO, BA64::ZERO),
+                is_trigger: Replicated::new(Boolean::FALSE, Boolean::FALSE),
+                breakdown_key: Replicated::new(BK::ZERO, BK::ZERO),
+                trigger_value: Replicated::new(TV::ZERO, TV::ZERO),
+                timestamp: Replicated::new(TS::ZERO, TS::ZERO),
+            };
+            padding_input_rows.push(row);
+        }
+    }
+
+    input.extend(padding_input_rows);
+    Ok(input)
+}
+
+/// # Errors
+/// Will propogate errors from `OPRFPaddingDp`
+/// # Panics
+///
+pub fn two_parties_add_dummies<C, BK, TV, TS, const B: usize>(
+    ctx: &C,
+    padding_input_rows: &mut Vec<OPRFIPAInputRow<BK, TV, TS>>,
+    h_i: Role,
+    h_i_plus_one: Role,
+    padding_params: &PaddingParameters,
+) -> Result<u32, Error>
+where
+    C: Context,
+    BK: BooleanArray + U128Conversions,
+    TV: BooleanArray,
+    TS: BooleanArray,
+{
+    let mut total_number_of_fake_rows = 0;
+    let (mut left, mut right) = ctx.prss_rng();
+    // The first is shared with the helper to the "left", the second is shared with the helper to the "right".
+    let mut rng = &mut right;
+    if ctx.role() == h_i {
+        rng = &mut right;
+    }
+    if ctx.role() == h_i_plus_one {
+        rng = &mut left;
+    }
```

Comment:
Can you use a match statement for this, and panic if ctx.role() is h_out?

Reply:
Not sure we can. h_out isn't in scope in this function.  I can add an extra assert that `assert!(h_i != h_i_plus_one);`